File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        char[] chars = s.toCharArray();
        HashMap<Character,Integer> map = new HashMap<>();
        int from = 0, size = 0;
        for (int i = 0; i < chars.length; i++) {
            if (map.containsKey(chars[i])) {
                from = Math.max(from, map.get(chars[i]) + 1);
            }
            size = Math.max(size, i - from + 1);
            map.put(chars[i], i);
        }
        return size;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        location_map = {}
        max_val, last_idx = 0, 0
        for i in range(len(s)):
            last_idx = max(last_idx, location_map.get(s[i], -1) + 1) 
            max_val = max(max_val, i - last_idx + 1)
            location_map[s[i]] = i
        return max_val